--- D:\国科大通知\体系结构实验\lab8_exp19_69\myCPU1\csr.v
+++ D:\国科大通知\体系结构实验\lab8_exp19_69\myCPU\csr.v
@@ -20,6 +20,7 @@
     output wire         has_int,
 
     input  wire         wb_ex_r,
+    input  wire [5:0]   wb_tlb_ex,
 
     //tlb
     input  wire[4 : 0] tlbop_bus_es, //tlbsrch,tlbrd,tlbwr,tlbfill,invtlb
@@ -46,6 +47,8 @@
 reg        csr_crmd_ie;
 reg        csr_crmd_pg;
 reg        csr_crmd_da;
+reg [1 :0] csr_crmd_datf;
+reg [1 :0] csr_crmd_datm;
 reg [1 :0] csr_prmd_pplv;
 reg        csr_prmd_pie;
 reg [12:0] csr_ecfg_lie;
@@ -73,7 +76,6 @@
 wire [31:0] coreid_in;
 wire        csr_ticlr_clr;
 
-wire [31:0] csr_crmd_rvalue;
 wire [31:0] csr_prmd_rvalue;
 wire [31:0] csr_ecfg_rvalue;
 wire [31:0] csr_esata_rvalue;
@@ -134,6 +136,8 @@
 
 wire         fake_ex;
 
+wire tlbr = wb_tlb_ex[0];
+
 //crmd
 always @(posedge clk) begin
     if (reset)
@@ -189,6 +193,16 @@
     end 
 end
 
+always @(posedge clk) begin
+    if(reset) begin
+        csr_crmd_datf <= 2'b0;
+        csr_crmd_datm <= 2'b0;
+    end else if(csr_we && csr_num == `CSR_CRMD) begin
+        csr_crmd_datf <= csr_wmask[`CSR_CRMD_DATF] & csr_wvalue[`CSR_CRMD_DATF] | ~csr_wmask[`CSR_CRMD_DATF] & csr_crmd_datf;
+        csr_crmd_datm <= csr_wmask[`CSR_CRMD_DATM] & csr_wvalue[`CSR_CRMD_DATM] | ~csr_wmask[`CSR_CRMD_DATM] & csr_crmd_datm;
+    end
+end
+
 //prmd
 always @(posedge clk) begin
     if (wb_ex_r) begin
@@ -244,7 +258,14 @@
 end
 
 //badv
-assign wb_ex_addr_err = (wb_ecode == `ECODE_ADE || wb_ecode == `ECODE_ALE);
+assign wb_ex_addr_err = wb_ecode == `ECODE_ADE 
+                    || wb_ecode == `ECODE_ALE
+                    || wb_ecode == `ECODE_TLBR
+                    || wb_ecode == `ECODE_PIF
+                    || wb_ecode == `ECODE_PIS
+                    || wb_ecode == `ECODE_PIL
+                    || wb_ecode == `ECODE_PPI
+                    || wb_ecode == `ECODE_PME;
 always @(posedge clk) begin
     if(wb_ex_r && wb_ex_addr_err)
         csr_badv_vaddr <= (wb_ecode == `ECODE_ADE && wb_esubcode == `ESUBCODE_ADEF) ? wb_pc : wb_vaddr;
@@ -548,7 +569,7 @@
 end
 
 //rvalue
-assign csr_crmd_rvalue   = {27'b0, csr_crmd_pg, csr_crmd_da, csr_crmd_ie, csr_crmd_plv};
+assign csr_crmd_rvalue   = {23'b0, csr_crmd_datm, csr_crmd_datf, csr_crmd_pg, csr_crmd_da, csr_crmd_ie, csr_crmd_plv};
 assign csr_prmd_rvalue   = {29'b0, csr_prmd_pie, csr_prmd_pplv};
 assign csr_ecfg_rvalue   = {19'b0, csr_ecfg_lie};
 assign csr_esata_rvalue  = {1'b0, csr_estat_esubcode, csr_estat_ecode, 3'b0, csr_estat_is};
@@ -574,7 +595,9 @@
 
 assign fake_ex = !wb_ex_r & wb_ex;
 
-assign csr_pc = ertn_flush ? csr_era_pc : wb_ex_r ? csr_eentry_rvalue : fake_ex ? wb_pc : 32'h00000000;
+assign csr_pc = ertn_flush ? csr_era_pc :
+                wb_ex_r ? (wb_ecode == 6'h3f ? csr_tlbrentry_rvalue : csr_eentry_rvalue) :
+                fake_ex ? wb_pc : 32'h00000000;
 
 //read data
 assign csr_rvalue = (csr_num == `CSR_CRMD) ? csr_crmd_rvalue :

