--- D:\国科大通知\体系结构实验\lab8_exp19_69\myCPU1\mycpu.vh
+++ D:\国科大通知\体系结构实验\lab8_exp19_69\myCPU\mycpu.vh
@@ -1,8 +1,8 @@
     `define BR_DATA_WD           34
-    `define FS_TO_DS_DATA_WD     65+1
-    `define DS_TO_ES_DATA_WD     192+10+1
-    `define ES_TO_MS_DATA_WD     235+5+1
-    `define MS_TO_WS_DATA_WD     226+5+1
+    `define FS_TO_DS_DATA_WD     65+1+6//6 for tlb_ex_bus
+    `define DS_TO_ES_DATA_WD     192+10+1+6
+    `define ES_TO_MS_DATA_WD     235+5+1+6+1
+    `define MS_TO_WS_DATA_WD     226+5+1+6+1
     `define WS_TO_RF_DATA_WD     41
     `define ES_FWD_BLK_DATA_WD   42
     `define MS_FWD_BLK_DATA_WD   42
@@ -73,6 +73,8 @@
     `define CSR_CRMD_PIE      2
     `define CSR_CRMD_DA       3
     `define CSR_CRMD_PG       4 
+    `define CSR_CRMD_DATF     6 :5
+    `define CSR_CRMD_DATM     8 :7
     `define CSR_PRMD_PPLV     1 :0
     `define CSR_PRMD_PIE      2
     `define CSR_ECFG_LIE      12:0
@@ -122,3 +124,5 @@
     `define ECODE_TLBR    6'h3f 
     `define ESUBCODE_ADEF 9'h0  
     `define ESUBCODE_ADEM 9'h1
+
+    `define ADDR_MAX 32'h80000000

